partitionmanager/cli.py

```diff
@@ -0,0 +1,440 @@
+"""
+Interface for running the partition manager from a CLI.
+"""
+
+from datetime import datetime, timedelta, timezone
+from pathlib import Path
+import argparse
+import logging
+import traceback
+import yaml
+
+import partitionmanager.bootstrap
+import partitionmanager.table_append_partition as pm_tap
+import partitionmanager.types
+import partitionmanager.stats
+import partitionmanager.sql
+
+PARSER = argparse.ArgumentParser(
+    description="""
+    In already-partitioned tables with an auto_increment key as the partition,
+    add a new partition at the current auto_increment value.
+"""
+)
+
+PARSER.add_argument(
+    "--log-level",
+    default=logging.INFO,
+    type=lambda x: getattr(logging, x.upper()),
+    help="Configure the logging level.",
+)
+PARSER.add_argument(
+    "--prometheus-stats", type=Path, help="Path to produce a prometheus statistics file"
+)
+PARSER.add_argument(
+    "--config", "-c", type=argparse.FileType("r"), help="Configuration YAML"
+)
+
+GROUP = PARSER.add_mutually_exclusive_group()
+GROUP.add_argument("--mariadb", help="Path to mariadb command")
+GROUP.add_argument(
+    "--dburl",
+    type=partitionmanager.types.toSqlUrl,
+    help="DB connection url, such as sql://user:pass@10.0.0.1:3306/database",
+)
+
+
+class Config:
+    """Configuration data that the rest of the tooling uses.
+
+    Can be created from both an argparse object of command-line arguments, from
+    a YAML file, both, and potentially be modified via unit tests.
+    """
+
+    def __init__(self):
+        self.tables = set()
+        self.dbcmd = None
+        self.noop = False
+        self.num_empty = 2
+        self.curtime = datetime.now(tz=timezone.utc)
+        self.partition_period = timedelta(days=30)
+        self.prometheus_stats_path = None
+
+    def from_argparse(self, args):
+        """Populate this config from an argparse result.
+
+        Overwrites only what is set by argparse.
+        """
+        if args.table:
+            for n in args.table:
+                self.tables.add(partitionmanager.types.Table(n))
+        if args.dburl:
+            self.dbcmd = partitionmanager.sql.IntegratedDatabaseCommand(args.dburl)
+        elif args.mariadb:
+            self.dbcmd = partitionmanager.sql.SubprocessDatabaseCommand(args.mariadb)
+        if "days" in args and args.days:
+            self.partition_period = timedelta(days=args.days)
+            if self.partition_period <= timedelta():
+                raise ValueError("Negative lifespan is not allowed")
+        if "noop" in args:
+            self.noop = args.noop
+        if "prometheus_stats" in args:
+            self.prometheus_stats_path = args.prometheus_stats
```

Comment:
Reasonable, though I think the likelihood of re-running previous commands is equally high either way, as then one is mutating multiple yaml files rather than CLI arguments.

Still, having only a single source of operators is always nice.

Reply:
I'll tackle this in #17